src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java

```diff
@@ -0,0 +1,408 @@
+     * @param name logical name for this data source (may be null)
+     * @param queryLookaheadBases look ahead this many bases during queries that produce cache misses
+     * @param targetFeatureType When searching for a {@link FeatureCodec} for this data source, restrict the search to codecs
+     *                          that produce this type of Feature. May be null, which results in an unrestricted search.
+     */
+    public void addFeatureDataSource(final String featurePath, final String name, final int queryLookaheadBases, final Class<? extends Feature> targetFeatureType) {
+        addFeatureDataSource(new FeatureInput<>(featurePath, name != null ? name : featurePath), queryLookaheadBases, targetFeatureType);
+    }
+
+    /**
+     * Creates a FeatureDataSource backed by the provided FeatureInput. We will look ahead the specified number of bases
+     * during queries that produce cache misses.
+     *
+     * @param featureInput a FeatureInput specifying a source of Features
+     * @param queryLookaheadBases look ahead this many bases during queries that produce cache misses
+     * @param targetFeatureType When searching for a {@link FeatureCodec} for this data source, restrict the search to codecs
+     *                          that produce this type of Feature. May be null, which results in an unrestricted search.
+     */
+    public void addFeatureDataSource(final FeatureInput<VariantContext> featureInput, final int queryLookaheadBases, final Class<? extends Feature> targetFeatureType) {
+        Utils.validateArg( queryLookaheadBases >= 0, "Query lookahead bases must be >= 0");
+
+        featureDataSources.add(new FeatureDataSource<>(featureInput, queryLookaheadBases, targetFeatureType));
+        invalidateCachedHeaderAndDictionary();
+    }
+
+    /**
+     * Returns the aggregate sequence dictionary for this source of Variants. Uses the dictionary resulting
+     * from merging the individual VCF headers (if present) for variant inputs, otherwise attempts to create a
+     * sequence dictionary from an index file (if present).
+     *
+     * @return the sequence dictionary derived from the input sources, or null if no dictionary could be created
+     * from either the header or an index file.
+     */
+    public SAMSequenceDictionary getSequenceDictionary() {
+        validateFeatureSources();
+        if (mergedDictionary == null && null != getHeader()) {
+            validateAllSequenceDictionaries();
+            final VCFHeader header = getHeader();
+            mergedDictionary = header.getSequenceDictionary();
+            if (mergedDictionary == null || mergedDictionary.isEmpty()) {
+                // last resort - try to get a sequence dictionary from one of the feature data sources (which
+                // will create one from the index if necessary)
+                for (FeatureDataSource<VariantContext> fds : featureDataSources) {
```

Comment:
The dictionary created from the index will only reflect the variants in that particular `FeatureDataSource`, and may not be usable with the variants from the other data sources, so this fallback may not be a good idea.